### ISSUE TITLE
Allow passing arguments to interactive window start-up script, including Visual Studio macros

### ISSUE THREAD

**Issue by shift-evgeny** [Microsoft/PTVS#1193]
In configuring an interactive environment with a start-up script there is no way to pass arguments to the start-up script. Please add another input textbox for this.

It should support Visual Studio macros like $(SolutionDir) and $(BinDir), same as Visual Studio External Tools configuration. This would allow the start-up script to reference the binaries of the currently open project (and similar).

**Comment by zooba** [comment]
I've actually been working on some fairly significant changes to interactive window startup, so this would make sense to support as well. Rather than having a startup script, we're looking at having a (configurable per environment) directory of scripts that are all run at startup.

Would this reduce your need to provide parameters to the script, assuming you had separate directories for each? Or is that just going to force you to duplicate things in a way you'd rather avoid (i.e. you'd rather switch between Python versions within the script rather than having multiple scripts)?

What kind of parameters are you thinking of passing through?

We could very easily pass through some macros as environment variables - probably named like `VS_SOLUTION_DIR` and `VS_BIN_DIR` etc. - though it isn't necessarily possible to get all of them, and not everything that VS "typically" uses (for C#, etc.) make sense for Python.

Here's a list of "easy" ones I can think of that may be useful:
* solution dir/name
* project dir/name (if we have one project; also set the CWD to this dir)
* startup file (if exactly one file/project is selected for startup)
* search paths (if we have one project, and pass them through in the environment's `PYTHONPATH` equivalent)
* environment variables in project debug settings (if we have one project)

There are probably more, but those are the ones I can think of right now. As environment variables, you'll be able to do whatever you like with them from a startup script.

**Comment by shift-evgeny** [comment]
Good to know that this is being improved!

I think being able to run multiple scripts at start-up is useful in general (as long as the order is deterministic, e.g. alphabetical), but it wouldn't help me to pass the current project's binaries directory to the script. I was thinking of passing $(BinDir), so that I can use that with clr.AddReferenceToFileAndPath() in the script.

Yes, having environment variable accessible in the script would be even better. Your list looks good , but I would add the output binary name (e.g. "MyProject.exe") for the start-up project. (I'm happy to hardcode that in my case.)

CWD should probably be set to the project output dir (not project source dir), like it is when launching the project from VS (if "Working directory" is left empty in the Debug settings).

**Comment by shift-evgeny** [comment]
By the way, as a workaround, I was trying to use the path of the start-up script itself, but I can't figure out any way to get it. `__file__` just returns `<module>`.

**Comment by zooba** [comment]
> as long as the order is deterministic, e.g. alphabetical

Will do.

> I was trying to use the path of the start-up script itself, but I can't figure out any way to get it inside the script

There currently isn't one, since (IIRC) we just read the code and `exec` it. I'll make sure that we get that right this time.

>  I would add the output binary name (e.g. "MyProject.exe") for the start-up project

Ah, are you using the interactive window with a C# project? Or a Python project? I'm guessing the former, since Python projects don't really have an output binary or directory (except where we need to fake one to trick some of the Azure tools).

**Comment by shift-evgeny** [comment]
Yes, it's a C# project (many projects, actually) and I'd like to load the binaries for use in a REPL shell.

I looked at `visualstudio_py_repl.py` and saw that run_file_as_main() sets `self.exec_mod.__file__ = filename` and then runs `code.Execute(self.exec_mod)`, but I don't know enough about IronPython to know if there's a way to get that `exec_mod` value inside the script somehow.
